### ISSUE TITLE
[HOLD for payment 2024-07-24] [$500]  Chat - Bad performance when opening, closing or zooming big image

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#38180]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.51-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**  https://expensify.testrail.io/index.php?/tests/view/4418462
**Email or phone of affected tester (no customers):**  vdargentotest+web031224@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
Pre-requisite: the user must be logged in
1. Go to any chat
2. Send the attached GIF
3. Send the attached image
4. Click on the image preview to open it
5. Verify the opening process is not smooth
6. Click on the image to zoom in
7. Verify is not instantly zoomed in, the behavior is slowly
8. Try to move the zoom view, verify is not smooth
9. Close the image
10. Verify the closing process is not smooth

## Expected Result:
There should be no performance issues when opening, zooming or closing big images

## Actual Result:
The behavior of the app gets slower when opening, zooming or closing the big image

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/bd11fa5f-47c6-4ba4-b4be-4c83e3fd4701

![Bug6411486_1710273547556!Big_image (1)](https://github.com/Expensify/App/assets/78819774/66b91727-3f02-456a-bc34-f2f0f80622bb)

![Bug6411486_1710276381171!typing-fast-typing](https://github.com/Expensify/App/assets/78819774/487649b1-53ed-4aed-b090-4b8b11802816)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0184b3ab8ca5055856</li>
        <li>Upwork Job ID: 1777877652768452608</li>
        <li>Last Price Increase: 2024-06-27</li>
<li>Automatic offers: </li>
<ul>
<li>wildan-m | Contributor | 102956097</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sobitneupane</details>

**Comment by lanitochka17** [comment]
@mallenexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp 
CC @quinthar

**Comment by mallenexpensify** [comment]
Check on in vip-vsb
https://expensify.slack.com/archives/C066HJM2CAZ/p1710294208664919

**Comment by mallenexpensify** [comment]
Adding to #vip-vsb since it's a bug that doesn't involve money.

**Comment by mallenexpensify** [comment]
Bumping to weekly while we await a status update

**Comment by mallenexpensify** [comment]
I think/hope this can be External

**Comment by sobitneupane** [comment]
@mallenexpensify I do agree. I think this can be handled externally.

**Comment by mallenexpensify** [comment]
@sobitneupane , can you review the above proposal plz?

**Comment by sobitneupane** [comment]
Thanks for the proposal @ShridharGoel 

We would want solution that solves the issue for big image as the one attached here. Thank you for your efforts

**Comment by sobitneupane** [comment]
@ShridharGoel Can you please add more details to RCA.

In which file does the issue lie? Where are you proposing to make the change? What will be the affect on quality of image after the change? Can you also attach what the output would look like?

**Comment by sobitneupane** [comment]
@ShridharGoel I am fine with slight delay when zooming big image rather than decreasing quality as it only happens first time I zoom the image after opening the modal. I believe we can improve the performance issue when opening and closing the modal without decreasing the quality.

@mallenexpensify What's your take?

If we are okay with decreasing the image resolution, rather than compressing it to half, I think we should calculate the aspect ratio to make either height or width equal to max allowed value.

**Comment by sobitneupane** [comment]
@ShridharGoel We can do [that](https://github.com/Expensify/App/issues/38180#issuecomment-2061171017).  But won't it degrade the time required to upload file. I mean once I chose a heavy file,how long will it take to open upload modal?

**Comment by mallenexpensify** [comment]
> What's your take?

Awaiting reply from @ShridharGoel on your comment above @sobitneupane .  If there's no degradation, that sounds like the way to go.   If there is, and we feel we might have to compromise, can you start a discussion in #expensify-open-source to get more 👀?  I know nothing about the coding behind the feature and fix so I don't feel comfortable making a guess or assumption.

**Comment by jsdev2547** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Chat - Bad performance when opening, closing or zooming big image.

### What is the root cause of that problem?
Opening big image:
The provided image is png which takes over 10 seconds to loads. The front end or backend should consider implementing support for webp which takes under 2 seconds to load.

Closing big image:
In AttachmentModal.tsx, the HeaderWithBackButton component has onCloseButtonPress={closeModal}, the onModalClose does not have implementation, the onModalHide can be used instead which calls Navigation.dismissModal() of ReportAttachments.tsx.
https://github.com/Expensify/App/blob/68e69623a8bca0dce5f384fa351d5cbb2214aeb4/src/components/AttachmentModal.tsx#L505

Zooming big image:
Zooming for web needs to be reviewed and improvised. The current implementation in useEffect is calling scrollableRef.current.scrollLeft = initialScrollLeft + moveX, which takes several milliseconds.
https://github.com/Expensify/App/blob/68e69623a8bca0dce5f384fa351d5cbb2214aeb4/src/components/ImageView/index.tsx#L165

### What changes do you think we should make in order to solve the problem?

For closing big image:
In AttachmentModal, onModalClose() has no effect due to lack of implementation, instead call onModalHide() which calls Navigation.dismissModal.
https://github.com/Expensify/App/blob/68e69623a8bca0dce5f384fa351d5cbb2214aeb4/src/components/AttachmentModal.tsx#L505

For opening big image:
Front end is capable of providing webp support by converting the image to webp using canvas api i.e canvas.toDataURL('image/webp'), backend can also convert provided image files (png, jpegs) into webp. Webp are much smaller in size than pngs and loads much faster than pngs and also do not decrease the image quality. Need to discuss.

Zooming big image:
Need to improve zooming using GestureDetector from 'react-native-gesture-handler'.


### What alternative solutions did you explore? (Optional)

**Comment by sobitneupane** [comment]
@jsdev2547 Similar to closing image, can we improve the opening image experience (without converting)? It is okay to show loading indicator by opening the attachment modal. But we have some lag while opening the image. What is the root cause for it?

**Comment by jsdev2547** [comment]
@sobitneupane I have revised proposal. I am finding the lag and UI becoming non-responsive when opening the big image is specific only to pngs. To test need to check for non-responsive behavior in UI, i.e. the attachment modal should be able to get closed when the image is loading. When testing, I found the delay in opening and closing the image is considerably more noticeable in pngs than in other formats. 
Also the zoom issue is also resolved when using backend provided 1024px jpg images for pngs. The advantage for using the 1024px jpg images is the conversion is already done by the backend.
Please do help test and confirm the findings if the delay is noticeable only in pngs then the provided 1024px jpg images could provide solution for the issue.

**Comment by sobitneupane** [comment]
> Also the zoom issue is also resolved when using backend provided 1024px jpg images for pngs.


@jsdev2547 I don't think your above statement is true. I can reproduce the issue even with 1024px jpg image (I am quite surprised tbh). I thought the issue was due to the dimension but the issue is being reproduced even with 1024px image. 

https://github.com/Expensify/App/assets/25876548/694d53e1-d400-46d8-b632-cdf6ca01cdd3

**Comment by jsdev2547** [comment]
> I can reproduce the issue even with 1024px jpg image

@sobitneupane When I reset the closing modal portion ie. remove setShouldLoadAttachment(false); from closeModal callback, then the closing lag occurs even with 1024px jpg image. To resolve the lag issue need to include setShouldLoadAttachment(false) in closeModal. I have prepared [branch](https://github.com/jsdev2547/App/tree/attachmentBigImage) for testing, try testing zooming on 1024px jpg. 

[poc.webm](https://github.com/Expensify/App/assets/161593779/f36c1576-b739-4af8-bbd4-91f41485edf9)

**Comment by sobitneupane** [comment]
Thanks for the update @jsdev2547. Your proposal will improve the experience but still the screen still freezes while opening the modal. The issue with zoom is still reproducible.

https://github.com/Expensify/App/assets/25876548/19cb5d4a-6eb2-40f1-bee9-6b2e62fbefd7

**Comment by jsdev2547** [comment]
@sobitneupane I wanted to confirm if the screen freeze and the zoom freeze are related to the StyleUtils.getZoomSizingStyle
https://github.com/Expensify/App/blob/64d22534b3a06fb516684198167ab2e263d99a0e/src/components/ImageView/index.tsx#L225
During testing, when I comment out the PressableWithoutFeedback component and leave only the Image component, the screen freeze disappears but so does the zoom. Perhaps the zoom functionality needs to be revised and use gesture handler technique i.e zoom using double tap gesture.

Please test the [branch](https://github.com/jsdev2547/App/tree/attachmentBigImage), to confirm if the screen freeze still exists. In the test branch, the image loading for png is using 1024x jpg, while the jpg are loading original size. The jpg take few seconds to load but the screen does freeze.

**Comment by sobitneupane** [comment]
@jsdev2547 My bad. I cannot reproduce the issue I mentioned here with this [changes](https://github.com/Expensify/App/commit/7305454ac68e682dd675543c68204f90ff3dfab6) (initial commit)

I have three major questions about your proposal though.
- I couldn't understand how does changing the format solve the issue? By making the change, are we referencing to image returned by backend instead of local image? In that case, why don't we use backend returned image instead of local image in the first place as we wait for the image to be uploaded before displaying it?
```
const [imageUrl, setImageUrl] = useState(() => {
    const { fileExtension } = FileUtils.splitExtensionFromFileName(url ?? '');
    const isvalid = (fileExtension === "png") ? true : false;
    const u = isvalid ? url + ".1024.jpg" : url;
    return u;
})
```
- Will the solution work for user uploading an image?
- There are the cases where we cannot change the url extension. For example: if we are referring to image that is not from Expensify. ```![https://hatrabbits.com/wp-content/uploads/2017/01/random.jpg](https://hatrabbits.com/wp-content/uploads/2017/01/random.jpg)```

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by sobitneupane** [comment]
Bump @jsdev2547 
https://github.com/Expensify/App/issues/38180#issuecomment-2084724901

**Comment by jsdev2547** [comment]
@sobitneupane 

> By making the change, are we referencing to image returned by backend instead of local image? In that case, why don't we use backend returned image instead of local image in the first place as we wait for the image to be uploaded before displaying it?
The backend returns the original uploaded image along with the 1024px jpeg versions for when the image is already uploaded.
When uploading attachment, currently the image preview is using the local image, the url is in blob format and the backend has yet to process the image and return the jpeg version. 

> There are the cases where we cannot change the url extension. For example: if we are referring to image that is not from Expensify. 

If the image is hosted externally, then the loading occurs without delay and even the big image loads without delay. So using the backend provided jpeg version has no impact.

> Will the solution work for user uploading an image?
The only problem remains for when uploading the attachment which uses the local image (blob url). The backend provided versions are only available after upload. I tried the manipulateAsync helper utility used by cropOrRotateImage which can convert images to jpeg, but the loading is much slower. The other approach, which I am still testing is using canvas api.

**Comment by jsdev2547** [comment]
@sobitneupane 

> By making the change, are we referencing to image returned by backend instead of local image? In that case, why don't we use backend returned image instead of local image in the first place as we wait for the image to be uploaded before displaying it?

The backend returns the original uploaded image along with the 1024px jpeg versions for when the image is already uploaded.
When uploading attachment, currently the image preview is using the local image, the url is in blob format and the backend has yet to process the image and return the jpeg version. 

> There are the cases where we cannot change the url extension. For example: if we are referring to image that is not from Expensify. 

If the image is hosted externally, then the loading occurs without delay and even the big image loads without delay. So using the backend provided jpeg version has no impact.

> Will the solution work for user uploading an image?

The only problem remains for when uploading the attachment which uses the local image (blob url). The backend provided versions are only available after upload. I tried the manipulateAsync helper utility used by cropOrRotateImage which can convert images to jpeg, but the loading is much slower. The other approach, which I am still testing is using canvas api.

**Comment by jsdev2547** [comment]
@sobitneupane I am still trying to find viable solution for when the user uploads the image attachment.

**Comment by jsdev2547** [comment]
@sobitneupane I am still trying to find reliable solution for when the user uploads the image attachment.

**Comment by mallenexpensify** [comment]
@sobitneupane , what do you propose are the next best steps here?

**Comment by sobitneupane** [comment]
We are still waiting for the better proposals.

@jsdev2547 Any update?

**Comment by sobitneupane** [comment]
Waiting on proposals

**Comment by mallenexpensify** [comment]
Thanks @sobitneupane , bumped to $500

**Comment by sobitneupane** [comment]
Waiting on Proposal

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Chat - Bad performance when opening, closing or zooming big image

### What is the root cause of that problem?

Image resolution too high to handle, particularly when interacting with the image through zoom or pan functions. By checking the performance tab in Chrome DevTools, we can see that the lag primarily occurs during the painting phase rather than the rendering phase.

<img width="565" alt="SCR-20240523-mllr" src="https://github.com/Expensify/App/assets/11847279/cf488d1f-ff8c-460b-8014-9e466d16afe4">

### What changes do you think we should make in order to solve the problem?

Simply don't show the image preview if it has to big resolution. This is the behavior of slack, they aren't shows the image if it's more than 8000px. But they provide download link.

<img width="896" alt="image" src="https://github.com/Expensify/App/assets/11847279/29797454-e40e-4e8a-832e-413b0fcf17c4">

We can add check in `src/components/Attachments/AttachmentView/index.tsx`
```tsx

    const isFileHaveDimension = (file: FileObject): file is ImagePickerResponse => {
        return 'width' in file && 'height' in file;
    };
    const isResolutionValid = file && isFileHaveDimension(file) && (file?.height ?? 0) <= 8000 && (file?.width ?? 0) <= 8000;

    if ((isImage || (file?.name && Str.isImage(file.name))) && isResolutionValid) {
        if (imageError && (typeof fallbackSource === 'number' || typeof fallbackSource === 'function')) {
```

If we need the high resolution image directly download the image, then in `ImageRendered.tsx` change `onPress` to:

```tsx
                <PressableWithoutFocus
                    style={[styles.noOutline]}
                    onPress={() => {
                        if ((imageWidth ?? 0) > 8000 || (imageHeight ?? 0) > 8000) {
                            const sourceURLWithAuth = addEncryptedAuthTokenToURL(source);
                            fileDownload(sourceURLWithAuth, imageName);
                            return;
                        }

```

[Branch to test.](https://github.com/Expensify/App/compare/main...wildan-m:App:wildan/fix/38180-dont-preview-big-image?expand=1)


### What alternative solutions did you explore? (Optional) 
We have similar performance issue with [avatar image picker](https://github.com/Expensify/App/issues/13006). We decided to validate the resolution.

If we want this approach, we can adopt it by modifying `isValidFile` and `validateAndDisplayFileToUpload` function in `src/components/AttachmentModal.tsx`

```tsx
       const isValidFile = useCallback((fileObject: FileObject) => {
            return new Promise<void>((resolve, reject) => {
                if (fileObject.size && fileObject.size > CONST.API_ATTACHMENT_VALIDATIONS.MAX_SIZE) {
                    setIsAttachmentInvalid(true);
                    setAttachmentInvalidReasonTitle('attachmentPicker.attachmentTooLarge');
                    setAttachmentInvalidReason('attachmentPicker.sizeExceeded');
                    reject();
                }

                if (fileObject.size && fileObject.size < CONST.API_ATTACHMENT_VALIDATIONS.MIN_SIZE) {
                    setIsAttachmentInvalid(true);
                    setAttachmentInvalidReasonTitle('attachmentPicker.attachmentTooSmall');
                    setAttachmentInvalidReason('attachmentPicker.sizeNotMet');
                    reject();
                }

                type AttachmentPickerType = typeof CONST.ATTACHMENT_PICKER_TYPE[keyof typeof CONST.ATTACHMENT_PICKER_TYPE];
                let fileType: AttachmentPickerType = CONST.ATTACHMENT_PICKER_TYPE.FILE;

                if (fileObject.type && fileObject.type.startsWith('image/')) {
                    fileType = CONST.ATTACHMENT_PICKER_TYPE.IMAGE;
                }

                if (fileType === CONST.ATTACHMENT_FILE_TYPE.IMAGE) {
                    isValidResolution(fileObject)
                        .then((isValid) => {
                            if (!isValid) {
                                setIsAttachmentInvalid(true);
                                setAttachmentInvalidReasonTitle('attachmentPicker.attachmentResolutionTooLarge');
                                setAttachmentInvalidReason('attachmentPicker.resolutionSizeExceeded');
                                reject();
                            } else {
                                resolve();
                            }
                        })
                        .catch(reject);
                } else {
                    resolve();
                }
            });
        }, []);
```

[Branch to test](https://github.com/Expensify/App/compare/main...wildan-m:App:wildan/fix/38180-validate-resolution?expand=1)

**Comment by sobitneupane** [comment]
I will review the proposal by EOD.

**Comment by sobitneupane** [comment]
Thanks for the proposal @wildan-m 

@wildan-m Backend returns the image in 1024px in jpg format which doesn't have any lag. I believe we can make use of it after the server returns the image in 1024 px. The issue only lies when user is uploading. What do you think of this approach?

**Comment by wildan-m** [comment]
@sobitneupane I believe that using 1024px for a thumbnail may not result in good quality when zooming in, is that for thumbnail purpose? also convert it to jpg will remove transparent layer of png image. If we want to use convert approach then to avoid lag during upload commit, we should consider having an API that can convert the image before it is previewed and committed.

**Comment by sobitneupane** [comment]
Thanks for the update @wildan-m 

> I believe that using 1024px for preview may not result in good quality when zooming in, is that for thumbnail purpose? also convert it to jpg will remove transparent layer of png image.

It's a good point.

@wildan-m Can you please attach screen recording(for your proposal) of the flow from uploading to previewing in chat report?

**Comment by wildan-m** [comment]
@sobitneupane proposal [updated](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141).

Add video demo and fix resolution check when uploading.

> Is there any specific reason for choosing 8000px as the cut off point? I think we should test it for lower resolution images as well to make sure we don't have any issue with lower resolution image.

I am simply replicating Slack behavior by conducting a quick test with various sample images. The lower the resolution, the quicker the process (approximate value). 8000px might be too high; we can decrease it if necessary.

**Comment by sobitneupane** [comment]
@wildan-m Thanks for the update.

We will consider your solution as an option, but my preference is to display the image if possible. I will wait for proposals over the weekend and start discussing your suggested approach early next week.

**Comment by mallenexpensify** [comment]
Checking on[ internally](https://expensify.slack.com/archives/C02NK2DQWUX/p1717436281241939).

**Comment by sobitneupane** [comment]
Being discussed internally.

**Comment by sobitneupane** [comment]
Started discussion in slack:
https://expensify.slack.com/archives/C066HJM2CAZ/p1718003479823739

**Comment by Julesssss** [comment]
> @wildan-m Backend returns the image in 1024px in jpg format which doesn't have any lag. I believe we can make use of it after the server returns the image in 1024 px. The issue only lies when user is uploading. What do you think of this approach?

Yeah this is correct. One complication on the backend was that very wide or tall images led to a very blurry preview image, so I recently started cropping these images. We should be careful to test an image like this one: 


`1024x10573`:
![nEyfOgA](https://github.com/Expensify/App/assets/10736861/a7c75a3b-fc69-4993-9c9e-02d6a87296ba)

**Comment by Julesssss** [comment]
Not overdue.

**Comment by Julesssss** [comment]
Lets double bounty

**Comment by sobitneupane** [comment]
Waiting for proposals

**Comment by wildan-m** [comment]
@sobitneupane How is [the discussion](https://github.com/Expensify/App/issues/38180#issuecomment-2157510979) going?

**Comment by sobitneupane** [comment]
@wildan-m [Here](https://github.com/Expensify/App/issues/38180#issuecomment-2158157452) is the input from @Julesssss

**Comment by wildan-m** [comment]
@sobitneupane It appears that @Julesssss has commented on [your feedback](https://github.com/Expensify/App/issues/38180#issuecomment-2134501212). My suggestion was not to optimize performance, but rather [to modify the behavior](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141) when uploading high-resolution images.

Even if we manage to enhance performance successfully, there should still be a limit on the resolution that the image preview feature can handle. Without constraints, it may result in additional lag or unexpected errors.

**Comment by Julesssss** [comment]
> Even if we manage to enhance performance successfully, there should still be a limit on the resolution that the image preview feature can handle

To clarify, I think we might be able to use the 1024x1024 thumbnail for the preview. 

But to be honest, I'm a bit lost about the proposals' state and which are still valid. If you have a proposal could you please share a sentence about the plan, and a link to the original proposal? Thanks

**Comment by NJ-2020** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Bad performance when opening, closing, or zooming a big image
### What is the root cause of that problem?
The image resolution pixel is too big 9000x9000 pixels.
When I test with simple HTML and local image (9000x9000) it's still rendering slowly when I zoom in and zoom out

**Test Result:**

https://github.com/Expensify/App/assets/76243292/a44d5c16-67d8-42de-b22d-7da6aa0ed572

### What changes do you think we should make in order to solve the problem?
We can detect if the pixels of the image are equal or more than 3000, we can resize the image using Sharp or if we don't want to use Sharp for react native, we can use another package.  But the main point is we need to resize the image to a smaller size for instance 1000 pixels or less

I've tested it with a local HTML file using Sharp and it renders the image very fast when I zoom in and zoom out in the browser

**Test Result:**

https://github.com/Expensify/App/assets/76243292/8519678f-79ff-4e47-a7aa-e42197d25b35


### What alternative solutions did you explore? (Optional)
We can set a limited number of pixels of the image when we're going to upload an attachment

**Comment by wildan-m** [comment]
@Julesssss this is the link to the [original proposal](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141).

**Comment by NJ-2020** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/38180#issuecomment-2186160693)

**Comment by Julesssss** [comment]
Okay thanks all. I still like @wildan-m's proposal because it allows us to make further backend improvements later on while still fixing the issue.

@ShridharGoel This works, but I would prefer to do compression on the backend, using our existing logic so that we don't have different resizing logic throughout the product. However, we don't have time for the backend work currently.

@NJ-2020 again, I'd prefer not to use a client library because this splits our image resizing logic into different solutions and it becomes messy.

**Comment by Julesssss** [comment]
@sobitneupane do you agree? 

One thing we'll need to improve is how the image is downloaded. @wildan-m in your example we immediately download the image when selected. But let's see if the design team can help us with how this should work.

**Comment by wildan-m** [comment]
For comparison, Slack will display a placeholder image/icon if the resolution is over ~7000px.

![image](https://github.com/Expensify/App/assets/11847279/5b590861-ce71-45bb-bf70-9bfc1369f11b)

**Comment by dannymcclain** [comment]
I'm not totally sure I'm following... but am I understanding correctly that we can't display a smaller preview image in the attachment modal because we are uploading the local file and therefore can't create a smaller preview for it yet?

If that's the case... I guess we just update the "empty" state of the attachment modal?

Some random options:

![image](https://github.com/Expensify/App/assets/7516624/e91f40a7-0cc2-4354-a4a4-bfa87a59f7f1)
Big placeholder with file info

![image](https://github.com/Expensify/App/assets/7516624/cfc27758-bb7d-4317-95a2-241d4f84f385)
Slight tweak on our existing "file" display

![image](https://github.com/Expensify/App/assets/7516624/75273cf9-94d3-4dd6-9566-dce6e20ad923)
Basically do nothing

@Expensify/design any thoughts/ideas/preferences on this one?

**Comment by Julesssss** [comment]
> but am I understanding correctly that we can't display a smaller preview image in the attachment modal because we are uploading the local file and therefore can't create a smaller preview for it yet?

Ah not quite. I shouldn't have added that screenshot as it confuses things. 

We will always generate a 1024x1024 thumbnail, which gets displayed in the item row, and **could** be shown. We just cant render the full-resolution image.

**Comment by shawnborton** [comment]
Hmm if we have the 1024x1024 preview image, I would love to use that. Though I kind of think we might need to consider doubling that in size so the image generally looks good on bigger screens, as 1024 will look blurry on most modern laptops I think?

**Comment by Julesssss** [comment]
@shawnborton I agree that a higher-resolution preview would be a better solution, but that would require a new/updated API on the backend, and there's no way we can prioritise that work currently.

**Comment by dannymcclain** [comment]
Oh hmm, then why wouldn't we just render the image that we can render, and then the download button in the attachment modal would download the full resolution image? I feel like I'm not understanding something.

**Comment by Julesssss** [comment]
> why wouldn't we just render the image that we can render, and then the download button in the attachment modal would download the full resolution image?

Nah you understand correctly, ^ that sounds great. So we'd just re-use the existing Image preview modal?

**Comment by dannymcclain** [comment]
Yeah I think so! @shawnborton do you have anything else to add?

**Comment by shawnborton** [comment]
Nope, sounds like the idea is to use the generated thumbnail we have in the view modal if the image is too big, right?

**Comment by wildan-m** [comment]
> Nope, sounds like the idea is to use the generated thumbnail we have in the view modal if the image is too big, right?

@Julesssss @dannymcclain @shawnborton Is the solution scalable? Please also consider [this case](https://github.com/Expensify/App/issues/38180#issuecomment-2158157452), especially with very long height.

Imagine we upload an image with a greater height than that scenario, and we need to make a smaller version of it.

My questions are:

1. Should we maintain the aspect ratio? If we resize the height to, let's say, 2048px, the width may become too small.
2. If we create a cropped version of the thumbnail and display it in the viewer, it won't accurately represent the actual image.

In my opinion, Slack's approach is excellent. There's no need to show them in the viewer if it's too large; allow the user's OS image viewer to manage it after downloading.

**Comment by Julesssss** [comment]
After reviewing the current behaviour of being able to pan and zoom, I do agree the proposed solution is quite a downgrade. 

If we show the preview after a certain height/width is met, I think we'd need to make sure that (**A**) the height/width cutoff is reasonably high 4000-5000? and (**B**) that we make it clear to the users that the rendered image is just a preview of the full downloadable image.

**Comment by sobitneupane** [comment]
There are two cases that need to be addressed:
1. Preview before the attachment is uploaded (we won't have the compressed/cropped image from the server at this point).
2. Preview after the attachment is uploaded.

**Comment by sobitneupane** [comment]
Probable solutions so far:

1. Crop the high-dimension image before displaying it
2. Display the compressed image (this only works for the [2nd case](https://github.com/Expensify/App/issues/38180#issuecomment-2191190709))
3. Show a placeholder instead of the image (suggested [here](https://github.com/Expensify/App/issues/38180#issuecomment-2188984490))

**Comment by wildan-m** [comment]
I would vote for 3rd option with 2nd/3rd design

**Comment by wildan-m** [comment]
WhatsApp uses the second approach with additional converting to JPEG, as displayed in their image viewer and after the image is re-downloaded:
![WhatsApp Image 2024-06-26 at 16 16 34](https://github.com/Expensify/App/assets/11847279/5925292f-93d6-4060-b7d1-27b52a330747)

It is evident that the image is unusable.

**Comment by Julesssss** [comment]
I think our backend logic is fine, a 1024* preview and full size download is good. We just need to figure out how to handle [these ](https://github.com/Expensify/App/issues/38180#issuecomment-2191190709)two cases in the client. 

- I suggest that while the preview is loading we show the loading spinner (as we currently do within the chat)
- Then when clicked, we show the model with the full-resolution image
    - UNLESS the image is above a specific width/height, in which case we replace the preview with the below component within the model:
    
<img width="236" alt="Screenshot 2024-06-26 at 11 00 49" src="https://github.com/Expensify/App/assets/10736861/3af41fd9-2072-4756-b26d-9fcf4d58ca57">

**Comment by dannymcclain** [comment]
>Then when clicked, we show the model with the full-resolution image
>UNLESS the image is above a specific width/height, in which case we replace the preview with the below component within the model:

Personally I don't see any value in opening a modal just to display a little file component/lockup. If we're not going to show them a preview of the image, we might as well treat it like any other non-previewable file and only allow them to download (not open in the modal at all). IMO it would be better to show a low-res preview with a little message that says something like "This image has been resized for previewing. Download for full resolution." I just think it would be better to give the user _something_ to see, even if it's downsized, rather than nothing. I think the most common use case for uploading photos to a chat app is to be able to see them in the chat app. So we should do our best to make that happen (even if the image preview is less than ideal). But that's just my 2 cents! @shawnborton @dubielzyk-expensify do y'all have any thoughts to add?

**Comment by shawnborton** [comment]
That sounds great to me Danny, I agree with you.

**Comment by Julesssss** [comment]
> "This image has been resized for previewing. Download for full resolution."

Cool, that sounds great. As (I believe) this is new UI, would you mind creating a mock we can pass onto @wildan-m for implementation?

**Comment by dannymcclain** [comment]
I was thinking something like this, where we just reuse our little banner from archived chats:
![image](https://github.com/Expensify/App/assets/7516624/7f43dba6-0ab6-4914-9696-73119013340f)

We could do something more minimal/out of the way if we wanted. Maybe something like this:
![image](https://github.com/Expensify/App/assets/7516624/60b61aa8-1940-49a7-946c-f65d68b57e94)

@Expensify/design what do you think?

**Comment by shawnborton** [comment]
I like it, I think I like the second version so it takes up less space? We could consider linking the word Download too to the download action?

**Comment by dannymcclain** [comment]
>I like it, I think I like the second version so it takes up less space? We could consider linking the word Download too to the download action?

Sounds good to me. This then yeah? I don't think it's _totally_ necessary, but I don't really mind it either. Would probably be helpful for some.

![image](https://github.com/Expensify/App/assets/7516624/90c3315a-7b4f-485d-96cd-38065b18cb22)

**Comment by shawnborton** [comment]
Cool, yeah that was what I was thinking. I can see keeping things simple and not doing the link, since there is an obvious icon  in the top right.

**Comment by wildan-m** [comment]
@sobitneupane @Julesssss Check out my updated proposal [here](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141).

Change log:
- Moved my previous main solution to alternative 2.
- Added a new solution based on recent requirement changes (including demo and branch link).

In my view, the [suggested solution](https://github.com/Expensify/App/issues/38180#issuecomment-2192527048) is similar to my previous proposed solution, with some minor differences.

Previous solution: Suggested not showing a preview or using a placeholder icon, and providing the original link.
Current requirement change: Suggests showing a downsized image preview and providing the original link.

Since this may be considered a new feature, I believe further minor adjustments (refactoring, adding an icon, styling) should be done in the PR review phase.

**Comment by wildan-m** [comment]
@shawnborton @dannymcclain please review the draft outcome in [this video](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141).

**Comment by shawnborton** [comment]
Pretty close, but some notes:
![CleanShot 2024-06-27 at 08 31 33@2x](https://github.com/Expensify/App/assets/2319350/1d813b3d-7c0f-4db4-8b38-6bd6e52ade5e)

- We decided not to use the blue download link
- There seems to be a good bit of space below the warning text and the green button below
- The text should use our textSupporting color

But all of this can be sorted out in the PR review process.

**Comment by shawnborton** [comment]
Also @mallenexpensify why is the bounty for this one so high? Can we knock it back down to $250 like all other bugs?

**Comment by sobitneupane** [comment]
@wildan-m 

We won't have the compressed image available for [the 1st case](https://github.com/Expensify/App/issues/38180#issuecomment-2191190709). How are you handling the 1st case? I don't think compressing it in client side is good idea as it might take some time to compress the image.

> Preview before the attachment is uploaded (we won't have the compressed/cropped image from the server at this point).

We will be displaying 1024 px compressed image (used by thumbnail) for the 2nd case. Is it available in `AttachmentView`? If not, how will we pass it?

**Comment by wildan-m** [comment]
@sobitneupane @shawnborton @dannymcclain Can the preview image for high-resolution image be zoomable? I believe we can utilize a local blob image for the case before uploading, with a non-zoomable preview. This would eliminate the need to wait for the image to be compressed from the server.

**Comment by shawnborton** [comment]
I would think our preview for this should just work the same as our other image previews when uploading.

**Comment by wildan-m** [comment]
@sobitneupane @Julesssss Is there an image compression API that can be used before sending the attachment?

If so, the flow for the first case would be as follows:
1. User uploads the file
2. The file is sent to the `compressImage` API
3. The compressed image is used for image preview
4. The user sends the attachment, which will send the original image file.

**Comment by sobitneupane** [comment]
@wildan-m I think it would be overkill. For the case before the image is uploaded, I think we should go with [this approach](https://github.com/Expensify/App/issues/38180#issuecomment-2188984490).

**Comment by wildan-m** [comment]
> Also @mallenexpensify why is the bounty for this one so high? Can we knock it back down to $250 like all other bugs?

@shawnborton This issue is 3 months old, so I believe the current bounty is fair. 

Here is [my perspective on reducing the bounty](https://expensify.slack.com/archives/C01GTK53T8Q/p1706077339989679?thread_ts=1706059963.264749&cid=C01GTK53T8Q).

**Comment by shawnborton** [comment]
I think this is a super edge case bug that has impacted nearly 0 paying customers, and therefore I think we should place it at the smallest starting point possible.

**Comment by shawnborton** [comment]
Alternatively, we could just close this issue out and wait for an actual paying customer to report it first before making any adjustments here.

**Comment by Julesssss** [comment]
> why is the bounty for this one so high? Can we knock it back down to $250 like all other bugs?

We didn't have any proposals for a long time, so I followed the process and doubled the bounty. It seems a bit unfair to reduce this because detailed proposals were made based on the current bounty at that time. I agree the solution ended up being less difficult than a detailed performance improvement might be, but we did get the desired outcome. 

As for whether we drop these changes, I want to move forward as this issue was created to improve app performance -- a major problem. If we just wanted to improve image uploading I would agree with you that this wasn't essential 👍

**Comment by shawnborton** [comment]
Can you elaborate more on what you mean by major problem? From my point of view, it's hard to call something a major problem if we haven't had a single customer report this issue. Even in our internal usage of the product, I don't know that anyone has encountered this?

**Comment by Julesssss** [comment]
> I would think our preview for this should just work the same as our other image previews when uploading.

I agree with this. And @sobitneupane's comment [here](https://github.com/Expensify/App/issues/38180#issuecomment-2193911824). With these I think we are very close to a working proposal now.

**Comment by Julesssss** [comment]
> can you elaborate more on what you mean by major problem? From my point of view, it's hard to call something a major problem if we haven't had a single customer report this issue.

Performance, in general, is the major problem. We have had many experts working to improve app performance and are seeking even minor gains. Most of our performance gains haven't been individually raised as problematic by an internal or external user, but have been detected and handled by 3rd party teams under the 'new dot-quality' project (previously new dot-performance).

In this case we are close to a proposal that has been approved by the design team and noticeably reduces UI lag when viewing an uploaded image, so my vote is to implement it. But I'd happily defer to the project team.

**Comment by shawnborton** [comment]
> reduces UI lag when viewing an uploaded image

My understanding is that the proposals only address viewing a super large image, not just any image. Is that your understanding too?

But you haven't really addressed my comments at all. I'm curious if you think this is a major problem impacting our users or even our internal usage. Thoughts?

**Comment by Julesssss** [comment]
Yeah, I share the understanding. The improvement scales with the dimensions, which is noticable only as the images get really large. 

Also no, I don't think there is an argument for this improvement being a major problem, nor impacting users. Where I disagree, is that in the case of app performance, I believe it is better to make the pre-optimisation (just for performance, not generally).

**Comment by Julesssss** [comment]
Alright. Sorry to reduce this bounty, but as we're on the fence about this issue I'm reducing the bounty to $500. That gives us a bit more of a reason to implement the solution.

**Comment by mallenexpensify** [comment]
The reason the bounty was high was because the issue has been open for months and it's been a difficult bug to fix (over 100 comments here and multiple proposals).  [I addressed in #contributor-plus](https://expensify.slack.com/archives/C02NK2DQWUX/p1717436281241939) and [Sobit addressed in #vip-vsb to discuss options](https://expensify.slack.com/archives/C066HJM2CAZ/p1718003479823739).

**Comment by wildan-m** [comment]
@sobitneupane @shawnborton @Julesssss my [proposal updated](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141)

Change log:
- Ensure high resolution preview image is zoomable
- Resolve the 1st case ([ref](https://github.com/Expensify/App/issues/38180#issuecomment-2193942554))
- Update video result

**Comment by sobitneupane** [comment]
@wildan-m Can you please address following comments.

> We will be displaying 1024 px compressed image (used by thumbnail) for the 2nd case. Is it available in AttachmentView? If not, how will we pass it? ([ref](https://github.com/Expensify/App/issues/38180#issuecomment-2193911824))

> We decided not to use the blue download link ([ref](https://github.com/Expensify/App/issues/38180#issuecomment-2193905335))

We have decided to go with 2nd design from [this comment](https://github.com/Expensify/App/issues/38180#issuecomment-2192167500).

> To distinguish between uploaded and non-uploaded images, we can use the isUsedInAttachmentModal prop.

I'm a bit confused about `isUsedInAttachmentModal`. This is an existing prop, right? What use case is it currently serving? How does it distinguish between uploaded and non-uploaded images?

**Comment by wildan-m** [comment]
@sobitneupane Proposal [Updated](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141)

Change log:
change `isUsedInAttachmentModal` to `isUploaded`

>I'm a bit confused about isUsedInAttachmentModal. This is an existing prop, right? What use case is it currently serving? How does it distinguish between uploaded and non-uploaded images?
@sobitneupane `isUsedInAttachmentModal` is not accurate, it will also true for uploaded receipt image, so I change it with new prop called `isUploaded`


> We decided not to use the blue download link (https://github.com/Expensify/App/issues/38180#issuecomment-2193905335)
Does that mean the 'Download' text is clickable but with the same grey color, or do we only display unclickable text?

> We have decided to go with 2nd design from https://github.com/Expensify/App/issues/38180#issuecomment-2192167500.
What is the name of that (i) icon? can someone point me to similar existing text in a screen? @shawnborton @dannymcclain

**Comment by shawnborton** [comment]
> or do we only display unclickable text?

No need to display any kind of link here, we have the download icon in the top right

> What is the name of that (i) icon? can someone point me to similar existing text in a screen?

The icon is [here](https://github.com/Expensify/App/blob/main/assets/images/info.svg). I don't know that we have the icon + text as a component but it should be easy to make with our style helpers (flex, gap2, fontLabelSupporting, etc)

**Comment by wildan-m** [comment]
@sobitneupane @shawnborton Proposal [Updated](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141)

Change Log:
- Remove download link
- Add Info icon
- Explain where to extract thumbnail image

**Comment by sobitneupane** [comment]
Thanks for the update @wildan-m 

Overall [your proposal](https://github.com/Expensify/App/issues/38180#issuecomment-2126395141) looks good to me. We can work on the minor details in the PR.

🎀 👀 🎀 C+ Reviewed

**Comment by wildan-m** [comment]
@sobitneupane I need more time to re-test the solution and prevent regressions, as there are a few glitches on native that I am working to fix.

**Comment by wildan-m** [comment]
@sobitneupane The [PR is ready](https://github.com/Expensify/App/pull/44807) for your review. The translation in slack is not confirmed yet. 

https://expensify.slack.com/archives/C01GTK53T8Q/p1719953316322399

**Comment by Julesssss** [comment]
PR still lin review, I'm bumping it today

**Comment by Julesssss** [comment]
PR merged, awaiting deployment 🎉

**Comment by mallenexpensify** [comment]
@sobitneupane plz complete the BZ checklist above.

**Comment by mallenexpensify** [comment]
Contributor: @wildan-m paid $500 via Upwork
Contributor+: @sobitneupane owed $500 via NewDot

@sobitneupane plz complete the BZ checklist above.

**Comment by sobitneupane** [comment]
> BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

Rather than a bug, it is a performance issue.

**Comment by sobitneupane** [comment]
**Regression Test Proposal**

1. Go to any chat and upload an image with dimension less than 7000*7000 px.
2. Verify that the original image is displayed and can be zoomed.
3. Select Send button.
4. Open the newly uploaded image and verify that the original image is displayed and can be zoomed.
5. Go to any chat and upload an image with dimension more than 7000*7000 px.
6. Verify that the attachment modal opens up with filename and info text at the bottom of the modal.
7. Select Send button.
8. Verify that the image is displayed and can be zoomed and info text is displayed at the bottom of the image.

Do we agree 👍 or 👎

**Comment by Julesssss** [comment]
@sobitneupane I think that is good, but lets add a case for downloading the image.

**Comment by sobitneupane** [comment]
@Julesssss Updated.

**Comment by JmillsExpensify** [comment]
$500 approved for @sobitneupane

**Comment by mallenexpensify** [comment]
Thanks @sobitneupane, test case created

- https://github.com/Expensify/Expensify/issues/416544
